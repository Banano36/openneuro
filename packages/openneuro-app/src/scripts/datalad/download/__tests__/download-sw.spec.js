import { checkBrowserEnvironment, awaitRegistration } from '../download-sw.js'

describe('dataset/download - Service Worker method', () => {
  describe('checkBrowserEnvironment', () => {
    it('should throw an error for no service worker', () => {
      const mockWindow = {
        navigator: {},
      }
      expect(() =>
        checkBrowserEnvironment(mockWindow),
      ).toThrowErrorMatchingSnapshot()
    })
    it('should throw an error for no ReadableStream', () => {
      const mockWindow = {
        navigator: { serviceWorker: {} },
      }
      expect(() =>
        checkBrowserEnvironment(mockWindow),
      ).toThrowErrorMatchingSnapshot()
    })
    it('throws no errors if service worker and readable stream are available', () => {
      const mockWindow = {
        navigator: { serviceWorker: jest.fn() },
        ReadableStream: jest.fn(),
      }
      expect(() => checkBrowserEnvironment(mockWindow)).not.toThrowError()
    })
  })
  describe('awaitRegistration', () => {
    it('calls next if registration is active', done => {
      const mockRegistration = { active: true }
      return expect(() =>
        awaitRegistration(done, {})(mockRegistration),
      ).not.toThrowError()
    })
    it('waits for worker if installing', done => {
      const mockWindow = {
        navigator: {
          serviceWorker: {
            // Transition to active state immediately
            addEventListener: jest.fn((state, callback) =>
              callback({ target: { state: 'active' } }),
            ),
            removeEventListener: jest.fn(),
          },
        },
      }
      const mockRegistration = {
        active: false,
        installing: true,
        waiting: false,
      }

      // Passes in done as next()
      return expect(() => {
        awaitRegistration(done, mockWindow)(mockRegistration)
        // Check that listeners were actually used
        expect(
          mockWindow.navigator.serviceWorker.addEventListener,
        ).toHaveBeenCalledTimes(1)
        expect(
          mockWindow.navigator.serviceWorker.removeEventListener,
        ).toHaveBeenCalledTimes(1)
      }).not.toThrowError()
    })
  })
})
