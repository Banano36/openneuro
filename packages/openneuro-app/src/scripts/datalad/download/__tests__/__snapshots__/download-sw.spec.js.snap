// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataset/download - Service Worker method checkBrowserEnvironment should throw an error for no ReadableStream 1`] = `"Web streams are required to download. Try a recent version of Chrome or see the FAQ for how to enable these features on Firefox."`;

exports[`dataset/download - Service Worker method checkBrowserEnvironment should throw an error for no service worker 1`] = `"Your browser must support service workers to download. See the FAQ for supported browsers."`;
